nixos-container: support user-defined extra binds

wavewave opened the pull request:
in this pull request, I implemented a new feature of declarative nixos-container, user-defined extra binding directories, in addition to `/nix/store`, `/nix/var`...  
This function enables data communication between host and containers, and the host effectively works as nfs server and containers works as nfs clients. It is quite useful for developing multiple server deployment on a single machine.

A directory in the host can be bound to a directory in the container as specified in the `extraBinds` attribute, whose value is a list of binding specifications. Each binding specification has `host`, `container` and `isReadOnly` attributes. `host` is the host directory, `container` is the directory mounted in the container and `isReadOnly` is a boolean attribute specifying whether the directory in the container will be read-only or not. 

The following is some example container setup in `/etc/nixos/configuration.nix`

```
containers.foo = 
    { privateNetwork = true;
      hostAddress = "192.168.100.10";
      localAddress = "192.168.100.11";
      autoStart = true;
      extraBinds = [ { host="/home/wavewave/www";
                       container="/webroot";
                       isReadOnly = true; }
                     { host="/home/wavewave/temp";
                       container="/home/wavewave/temp";
                       isReadOnly = false; }
                   ];
      config =
        {config, pkgs, ...}:
        { networking.firewall.enable = false;
          services.openssh.enable = true;     
          services.lighttpd = {
            enable = true;
            document-root = "/webroot";
          };      
        };
    };
```

wavewave committed:
extraBindsRO/extraBindsRW

wavewave committed:
unify extraBindsRW/RO into extraBinds. Now arbitrary mount point is supported.

copumpkin commented:
This looks very appealing, thanks! :+1: 

This isn't actually using bind mounts, right? The filesystem namespacing takes care of the binding transparently?

lethalman commented:
Can we just have extraFlags? Because that's what it is.

edolstra commented:
Can you make it an attribute set (like e.g. `fileSystems`)? E.g.
```
bindMounts."/webroot" =
  { hostPath = "/home/wavewave/www";
    isReadOnly = true;
  };
```

edolstra commented:
In fact, maybe this should just piggy-back on `fileSystems`, e.g.
```
fileSystems."/webroot".hostPath = "/home/wavewave/www";
```

wavewave committed:
nixos-container: rename extraBinds to bindMounts and use attribute set format.

wavewave commented:
@edolstra I've renamed `extraBinds` to `bindMounts` and changed the syntax from a list to an attribute set like `fileSystems` as you suggested. Ideally, to incorporate this into `fileSystems` is a good idea, but it's too intrusive change for me as a nixos newbie to proceed ;-) 

@copumpkin This is using the `--bind` or `--bind-ro` option of `systemd-nspawn`, which I think uses `mount --bind` internally. I am not sure though.

@lethalman Using `extraFlags` was the very first idea I actually tried. At least, the current nixos code that deal with `extraFlags` need to be changed, and I think that this feature is quite good for generic purpose, so I separated it out. Still, additional support for user-defined `extraFlags` should be good for some expert usage.

wavewave commented:
Now in `/etc/nixos/configuration.nix`, one can specify `bindMounts` like the following.
Note that `hostPath` can be omitted if it is intended to be the same as the mounted path inside the container.
```
containers.foo = 
    { privateNetwork = true;
      hostAddress = "192.168.100.10";
      localAddress = "192.168.100.11";
      autoStart = true;
      bindMounts = { "/webroot" = { 
                       hostPath="/home/wavewave/www";
                       isReadOnly = true; 
                     };
                     "/home/wavewave/temp" = { 
                       isReadOnly = false; 
                    };
                  };
      config =
        {config, pkgs, ...}:
        { networking.firewall.enable = false;
          services.openssh.enable = true;     
          services.lighttpd = {
            enable = true;
            document-root = "/webroot";
          };      
        };
    };
```

edolstra commented on nixos/modules/virtualisation/containers.nix:
```diff
@@ -41,6 +41,41 @@ let
+        description = "Location of the host path to be mounted";
+      };
+      isReadOnly = mkOption {
+        default = false;
```
Maybe the default should be read-only, since that's the "safe" default?

edolstra commented on nixos/modules/virtualisation/containers.nix:
```diff
@@ -41,6 +41,41 @@ let
+      mountPoint = mkOption {
+        example = "/mnt/usb";
+        type = types.str;
+        description = "Location of the mounted in the container file systems";
```
"of the mounted"?

edolstra commented on nixos/modules/virtualisation/containers.nix:
```diff
@@ -334,6 +387,9 @@ in
              AUTO_START=1
            ''}
+
+           EXTRABINDS="${mkBindFlags cfg.bindMounts}"
```
This should be something like `$EXTRA_NSPAWN_FLAGS` as @lethalman suggested.

lethalman commented:
@wavewave why current code needs to be changed? Instead of having that length bindMounts option, just pass flags directly. Like you are doing with bindMounts, except without a special format.

edolstra commented:
@lethalman Having a dedicated, typed option is user-friendlier, and independent from the container implementation (i.e. we could replace systemd-nspawn with something else).

wavewave commented:
@lethalman currently `extraFlags` is not an option for a container afaik. So we need to make that as an option?
```
$  sudo nixos-rebuild switch -I nixpkgs=~/repo/srcc/nixpkgs 
building Nix...
building the system configuration...
error: The option `containers.foo.extraFlags' defined in `/etc/nixos/configuration.nix' does not exist.
(use ‘--show-trace’ to show detailed location information)
```

wavewave committed:
nixos-containers: bindMounts: change default to readOnly. use EXTRA_NSPAWN_FLAGS

wavewave commented:
@edolstra changed as you suggested in the code comments.

vcunat commented:
Ping.

edolstra commented on nixos/modules/virtualisation/containers.nix:
```diff
@@ -41,6 +41,41 @@ let
+      hostPath = mkOption {
+        default = null;
+        example = "/home/alice";
+        type = types.uniq (types.nullOr types.string);
```
This can be just `types.nullOr types.str`.

edolstra commented on nixos/modules/virtualisation/containers.nix:
```diff
@@ -41,6 +41,41 @@ let
+        default = null;
+        example = "/home/alice";
+        type = types.uniq (types.nullOr types.string);
+        description = "Location of the host path to be mounted";
```
Nitpick: descriptions should end with a period.

wavewave committed:
minor changes

wavewave committed:
Merge branch 'master' into extra-binds

wavewave commented:
I updated the branch to be compatible with the latest master. Please let me know if something is needed to be accepted.

lethalman commented:
I think it's ok, except this systemd template system is preventing yet another useful feature: we cannot `RequireMountsFor` the extra binds :(

anderspapitto commented:
@lethalman  do you mean
- "I consider this PR ready to merge, and in the future I would like see an addition that integrates better with RequireMountsFor"

or do you mean
- "There is an issue with RequireMountsFor that needs to be fixed before this PR can proceed"

?

I ask because I (and others, I guess) want this functionality to land. I'm about to cherry-pick this change to my local nixpkgs checkout but of course I'd prefer to have the "officially merged" version

ericsagnes commented:
Any update on this?

> I think it's ok, except this systemd template system is preventing yet another useful feature: we cannot RequireMountsFor the extra binds :(
 
It seems that #3021 implements a unit per container, could that help with this problem?

wavewave commented:
Is there anything for me to do to push this forward? I hope that this feature will land well...

lethalman commented:
@wavewave no let's just merge. Thanks.

copumpkin commented:
@wavewave does this help imperative containers at all or do those not follow this code path at all?
